driver-core/src/main/java/com/datastax/driver/core/UnresolvedUserType.java

```diff
@@ -0,0 +1,130 @@
+ *
+ *   Unless required by applicable law or agreed to in writing, software
+ *   distributed under the License is distributed on an "AS IS" BASIS,
+ *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *   See the License for the specific language governing permissions and
+ *   limitations under the License.
+ */
+package com.datastax.driver.core;
+
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicReference;
+
+import com.datastax.driver.core.exceptions.UnresolvedUserTypeException;
+
+import static com.datastax.driver.core.Metadata.escapeId;
+
+/**
+ * An unresolved user-defined type.
+ * Unresolved user-defined types can appear in the metadata when
+ * a column or a field references a user-defined type whose definition
+ * cannot be found in the current metadata.
+ *
+ * @since 3.0.0
+ */
+class UnresolvedUserType extends UserType {
+
+    private final AtomicReference<UserType> resolved = new AtomicReference<UserType>(null);
+
+    private final Metadata metadata;
+
+    UnresolvedUserType(String keyspaceName, String typeName, Cluster cluster) {
+        super(keyspaceName, typeName, Collections.<Field>emptyList(),
+            cluster.getConfiguration().getProtocolOptions().getProtocolVersion(),
+            cluster.getConfiguration().getCodecRegistry());
+        this.metadata = cluster.getMetadata();
+    }
+
+    @Override
+    public int size() {
+        resolve();
+        return resolved.get().size();
+    }
+
+    @Override
+    public boolean contains(String name) {
+        resolve();
+        return resolved.get().contains(name);
+    }
+
+    @Override
+    public Iterator<Field> iterator() {
+        resolve();
+        return resolved.get().iterator();
+    }
+
+    @Override
+    public Collection<String> getFieldNames() {
+        resolve();
+        return resolved.get().getFieldNames();
+    }
+
+    @Override
+    public DataType getFieldType(String name) {
+        resolve();
+        return resolved.get().getFieldType(name);
+    }
+
+    @Override
+    public String exportAsString() {
+        resolve();
+        return resolved.get().exportAsString();
+    }
+
+    @Override
+    public String asCQLQuery() {
+        resolve();
+        return resolved.get().asCQLQuery();
+    }
+
+    @Override
+    public int hashCode() {
+        resolve();
+        return resolved.get().hashCode();
+    }
+
+    @Override
+    Field[] getFields() {
+        resolve();
+        return resolved.get().getFields();
+    }
+
+    @Override
+    Map<String, int[]> getFieldIndicesByName() {
+        resolve();
+        return resolved.get().getFieldIndicesByName();
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (!(o instanceof UserType))
+            return false;
+        resolve();
+        return resolved.get().equals(o);
+    }
```

Comment:
There are a few issues here:

1. an instance is not equal to itself
1. an instance is not equal to another unresolved type referencing the same UDT
1. equality is not symetric (the actual UserType is not equal to an unresolved type that references it)
1. equal objects can have different hashcodes

1 and 2 are easily solved, but 3 and 4 are more complex. I'm not really sure if an unresolved type should be equal to the type it references.